# 0. Try approaches to find an effective way to merge datasets

In [5]:
# a trial case to load data into Python
import pandas as pd
from pandas.core.computation.check import NUMEXPR_INSTALLED

#file_path = ".././studies/TN_01/TrialNet_01_Data/TN01_ADVERSEEVENTS.CSV"
file_path = "/home/ec2-user/SageMaker/studies/TN_01/TrialNet_01_Data/TN01_nh01_screening.csv"

df1 = pd.read_csv(file_path)
df1.head()

/home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.0' or newer of 'numexpr' (version '2.7.3' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
/tmp/ipykernel_33999/705450557.py:8: DtypeWarning: Columns (46) have mixed types. Specify dtype option on import or set low_memory=False.
  df1 = pd.read_csv(file_path)


MaskID    Visit_Dt      Visit Visit_Type  ScreeningDay ScreeningMonth  \
0  200000  06/29/2007  Screening        NaN          29.0            Jun   
1  200003  06/29/2015  Screening        NaN          29.0            Jun   
2  200004  08/11/2015  Screening        NaN          11.0            Aug   
3  200007  11/21/2016  Screening        NaN          21.0            Nov   
4  200010  10/21/2016  Screening        NaN          21.0            Oct   

   ScreeningYear   Age RelativeWithT1D  DegreeOfRelative_1st  ...  \
0         2007.0  20.0             Yes                     0  ...   
1         2015.0  20.0             Yes                     1  ...   
2         2015.0  44.0             Yes                     1  ...   
3         2016.0   7.0             Yes                     1  ...   
4         2016.0  34.0             Yes                     1  ...   

   clinicalcenternumber         ScreeningSiteName ScreeningSiteNumber  \
0                   6.0  Ponce School of Medicine               903.0   
1                   NaN                       NaN              3017.0   
2                   NaN                       NaN              3126.0   
3                   NaN                       NaN              3126.0   
4                   NaN                       NaN                 NaN   

  WhichProtocolVersion  shipspecimencollectionkit  \
0                  NaN                        NaN   
1                  NaN                        NaN   
2                  NaN                        NaN   
3                  NaN                        NaN   
4                  NaN                        NaN   

  Specimencollectionkitshipmentd Spring2019ConsentDateMonth  \
0                            NaN                        NaN   
1                            NaN                        NaN   
2                            NaN                        NaN   
3                            NaN                        NaN   
4                            NaN                        NaN   

   selecttestkittype  Spring2019ConsentDateDay  Spring2019ConsentDateYear  
0                NaN                       NaN                        NaN  
1                NaN                       NaN                        NaN  
2                NaN                       NaN                        NaN  
3                NaN                       NaN                        NaN  
4                NaN                       NaN                        NaN  

[5 rows x 66 columns]

In [19]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234614 entries, 0 to 234613
Data columns (total 66 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   MaskID                            234614 non-null  int64  
 1   Visit_Dt                          234614 non-null  object 
 2   Visit                             234614 non-null  object 
 3   Visit_Type                        4674 non-null    object 
 4   ScreeningDay                      233579 non-null  float64
 5   ScreeningMonth                    233579 non-null  object 
 6   ScreeningYear                     233579 non-null  float64
 7   Age                               233440 non-null  float64
 8   RelativeWithT1D                   233554 non-null  object 
 9   DegreeOfRelative_1st              234614 non-null  int64  
 10  DegreeOfRelative_2ndor3rd         234614 non-null  int64  
 11  HaveDiabetes                      233488 non-null  o

In [20]:
df1.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234614 entries, 0 to 234613
Data columns (total 66 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   MaskID                            234614 non-null  int64  
 1   Visit_Dt                          234614 non-null  object 
 2   Visit                             234614 non-null  object 
 3   Visit_Type                        4674 non-null    object 
 4   ScreeningDay                      233579 non-null  float64
 5   ScreeningMonth                    233579 non-null  object 
 6   ScreeningYear                     233579 non-null  float64
 7   Age                               233440 non-null  float64
 8   RelativeWithT1D                   233554 non-null  object 
 9   DegreeOfRelative_1st              234614 non-null  int64  
 10  DegreeOfRelative_2ndor3rd         234614 non-null  int64  
 11  HaveDiabetes                      233488 non-null  o

# 1. Load merged datasets

## 1.1 TN_01: Merging into one

### 1.1.1 Data type transformation: string into category, float64 into float16

In [1]:
import pandas as pd
from pandas.core.computation.check import NUMEXPR_INSTALLED
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# The following code is to merge the two datasets for TN_01, but will cause error or dead kernel. 
# Convert the 'ProtocolVersion' column to string in both DataFrames
file_path_1 = "TN_01_merged_data_1.csv"
file_path_2 = "TN_01_merged_data_23.csv"

merged_first_part = pd.read_csv(file_path_1)
merged_data_23 = pd.read_csv(file_path_2)

/home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.0' or newer of 'numexpr' (version '2.7.3' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED


In [3]:
print(merged_first_part.info())
print()
print(merged_data_23.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 519750 entries, 0 to 519749
Columns: 1563 entries, MaskID to SignedConsentFor2017Protocol
dtypes: float64(665), int64(1), object(897)
memory usage: 6.1+ GB
None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1968807 entries, 0 to 1968806
Data columns (total 17 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   MaskID                         int64  
 1   Date_of_Registration           object 
 2   Status                         object 
 3   Date_of_Study_Start            object 
 4   Date_of_Draw                   object 
 5   Event_Title                    object 
 6   Spec_Name                      object 
 7   SampleMaskID                   object 
 8   Test_Name                      object 
 9   Result                         object 
 10  Result_Type                    object 
 11  Visit                          object 
 12  Date_at_Test_Results_Reported  object 
 13  Date_at_Ev

In [4]:
print(merged_first_part.memory_usage(deep=True).sum())
print()
print(merged_data_23.memory_usage(deep=True).sum())

17883089368

1802078130


In [5]:
import pandas as pd
import numpy as np

def transform_data_types(df):
    for column in df.columns:
        # Check for NaN values in the column
        has_nan = df[column].isnull().any()

        if df[column].dtype == 'float64':
            # Convert to int16 if all values are integers and there are no NaN values
            if not has_nan and all(df[column].dropna().apply(float.is_integer)):
                df[column] = df[column].astype('int16')
            else:
                # Convert to float16 otherwise
                df[column] = df[column].astype('float16')

        elif df[column].dtype == 'object':
            # Convert string columns to category
            if all(df[column].apply(lambda x: isinstance(x, str) or pd.isnull(x))):
                df[column] = df[column].astype('category')

    return df

In [6]:
transformed_merged_first_part = transform_data_types(merged_first_part)
transformed_merged_second_part = transform_data_types(merged_data_23)

print(transformed_merged_first_part.memory_usage(deep=True).sum())
print()
print(transformed_merged_second_part.memory_usage(deep=True).sum())

1311209169

245206343


In [7]:
print(transformed_merged_first_part.info())
print()
print(transformed_merged_second_part.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 519750 entries, 0 to 519749
Columns: 1563 entries, MaskID to SignedConsentFor2017Protocol
dtypes: category(888), float16(665), int64(1), object(9)
memory usage: 1.1+ GB
None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1968807 entries, 0 to 1968806
Data columns (total 17 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   MaskID                         int64   
 1   Date_of_Registration           category
 2   Status                         category
 3   Date_of_Study_Start            category
 4   Date_of_Draw                   category
 5   Event_Title                    category
 6   Spec_Name                      category
 7   SampleMaskID                   category
 8   Test_Name                      category
 9   Result                         object  
 10  Result_Type                    category
 11  Visit                          category
 12  Date_at_Test_Results_Report

### 1.2.2 Data merge into one

In [8]:
common_headers_123 = set.intersection(set(transformed_merged_first_part.columns), set(transformed_merged_second_part.columns))

# Merge the first two DataFrames on common headers using outer join
merged_data_all = pd.merge(transformed_merged_first_part, transformed_merged_second_part, on=list(common_headers_123), how='outer', 
                            suffixes=('_dup_x', '_dup_y'))

print(merged_data_all.info())
merged_data_all.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923848 entries, 0 to 2923847
Columns: 1578 entries, MaskID to LabID
dtypes: category(900), float16(666), int64(1), object(11)
memory usage: 6.4+ GB
None


MaskID    Visit_Dt Visit _22Jul2009VisitType  \
0  200023  07/01/2015   PRN                 NaN   
1  200023  07/01/2015   PRN                 NaN   
2  200023  07/01/2015   PRN                 NaN   
3  200071  01/24/2018   PRN                 NaN   
4  200071  01/14/2019   PRN                 NaN   

                     _15Aug2011VisitType StorageOfResidualsDuringTrialN  \
0  Rescreen/Confirmatory Screening Visit                            NaN   
1  Rescreen/Confirmatory Screening Visit                            NaN   
2  Rescreen/Confirmatory Screening Visit                            NaN   
3  Rescreen/Confirmatory Screening Visit                            NaN   
4                                    NaN                            NaN   

  _15Aug2011ConsentForResidualSa  _22Jul2009ConsentSignedDay  \
0                            Yes                         NaN   
1                            Yes                         NaN   
2                            Yes                         NaN   
3                            Yes                         NaN   
4                            NaN                         NaN   

  _22Jul2009ConsentSignedMonth  _22Jul2009ConsentSignedYear  ...  \
0                          NaN                          NaN  ...   
1                          NaN                          NaN  ...   
2                          NaN                          NaN  ...   
3                          NaN                          NaN  ...   
4                          NaN                          NaN  ...   

                Spec_Name SampleMaskID  Test_Name  Result Result_Type  \
0  Serum - Autoantibodies     S3827078     GAD65H       0        RPTD   
1  Serum - Autoantibodies     S3827078       MIAA       0        RPTD   
2  Serum - Autoantibodies     S3827078      IA-2H       0        RPTD   
3                     NaN          NaN        NaN     NaN         NaN   
4                     NaN          NaN        NaN     NaN         NaN   

   Date_at_Test_Results_Reported Date_at_Evaluation Date_Received  \
0                     07/20/2015         07/14/2015    07/14/2015   
1                     07/20/2015         07/14/2015    07/14/2015   
2                     07/20/2015         07/14/2015    07/14/2015   
3                            NaN                NaN           NaN   
4                            NaN                NaN           NaN   

  Date_Shipped  LabID  
0   07/08/2015  547.0  
1   07/08/2015  547.0  
2   07/08/2015  547.0  
3          NaN    NaN  
4          NaN    NaN  

[5 rows x 1578 columns]

#### The merged dataset is too large to be saved in the working directory. We are to turn columns with numerical values and large portion of missing values into sparse data structure.

In [9]:
summary_df = pd.read_csv('merged_data_summary.csv')
summary_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1578 entries, 0 to 1577
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   header              1578 non-null   object 
 1   Missing Percentage  1578 non-null   float64
 2   Value Type          1578 non-null   object 
 3   Top 1               1577 non-null   object 
 4   Top 2               1192 non-null   object 
 5   Top 3               708 non-null    object 
 6   Top 4               580 non-null    object 
 7   Top 5               526 non-null    object 
 8   Top 1 Value Count   1577 non-null   float64
 9   Top 2 Value Count   1192 non-null   float64
 10  Top 3 Value Count   708 non-null    float64
 11  Top 4 Value Count   580 non-null    float64
 12  Top 5 Value Count   526 non-null    float64
dtypes: float64(6), object(7)
memory usage: 160.4+ KB


### 1.2.4 Numerical data structure transformation

In [10]:
# Filter based on Missing Percentage and Value Type
filtered_columns = summary_df[(summary_df['Missing Percentage'] > 85) & 
                             (summary_df['Value Type'].isin(['float16', 'int64']))]

# Get the list of headers from the filtered DataFrame
headers_with_missing_85 = filtered_columns['header'].tolist()

# Print the count of columns
print("Count of columns with missing Percentage > 85% and Value Type 'float16' or 'int64':", len(headers_with_missing_85))

Count of columns with missing Percentage > 85% and Value Type 'float16' or 'int64': 635


#### Tried many times and found that: 1) for columns with very high missing values, sparse data structure significantly save memory; 2) when missing value less than 85%, sparse data structure does have advantage over float16. 

In [11]:
print(merged_data_all.info())
print()
print(merged_data_all.memory_usage(deep=True).sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923848 entries, 0 to 2923847
Columns: 1578 entries, MaskID to LabID
dtypes: category(900), float16(666), int64(1), object(11)
memory usage: 6.4+ GB
None

7855246123


In [12]:
error_columns = []  # List to store columns that raised errors

for col in headers_with_missing_85:
    try:
        print("Before transformation of", col, ":", merged_data_all[col].memory_usage(deep=True), "bytes")
        
        # Check data type of the column
        col_dtype = merged_data_all[col].dtype

        if col_dtype == 'float16' or col_dtype == 'int64':
            # Convert numeric columns to default sparse type
            merged_data_all[col] = pd.to_numeric(merged_data_all[col], errors='coerce').astype('Sparse[float16]')
        # No need to replace the column in the original DataFrame
        # Just printing out the memory usage after transformation
        print("After transformation to Sparse of", col, ":", merged_data_all[col].memory_usage(deep=True), "bytes")
    except ValueError as e:
        print(f"Error converting column {col}: {e}")
        error_columns.append(col)

    print()

# Print out columns that had errors
if error_columns:
    print("Columns with errors:", error_columns)

Before transformation of Thiazolidinediones_dup22 : 5847824 bytes
After transformation to Sparse of Thiazolidinediones_dup22 : 128 bytes

Before transformation of _2cTN18Exclusion_1aAbnormalGluc1 : 5847824 bytes
After transformation to Sparse of _2cTN18Exclusion_1aAbnormalGluc1 : 140 bytes

Before transformation of participantnoteligib_5Forfemales : 5847824 bytes
After transformation to Sparse of participantnoteligib_5Forfemales : 140 bytes

Before transformation of TN22ExclusionCriteri_Historyoftr : 5847824 bytes
After transformation to Sparse of TN22ExclusionCriteri_Historyoftr : 140 bytes

Before transformation of TN22ExclusionCriteri_Historyofre : 5847824 bytes
After transformation to Sparse of TN22ExclusionCriteri_Historyofre : 140 bytes

Before transformation of _2cTN20Exclusion_1DiagnosedwithD : 5847824 bytes
After transformation to Sparse of _2cTN20Exclusion_1DiagnosedwithD : 140 bytes

Before transformation of _2cTN20Exclusion_8Ongoinguseofme : 5847824 bytes
After transformati

In [13]:
print(merged_data_all.info())
print()
print(merged_data_all.memory_usage(deep=True).sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923848 entries, 0 to 2923847
Columns: 1578 entries, MaskID to LabID
dtypes: Sparse[float64, nan](635), category(900), float16(31), int64(1), object(11)
memory usage: 3.0+ GB
None

4248338911


In [14]:
for column in merged_data_all.columns:
    if merged_data_all[column].dtype == 'object':
        merged_data_all[column] = merged_data_all[column].astype('category')

In [15]:
#merged_data_all = merged_data_all.drop('Unnamed: 0', axis=1)

In [16]:
print(merged_data_all.info())
print()
print(merged_data_all.memory_usage(deep=True).sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923848 entries, 0 to 2923847
Columns: 1578 entries, MaskID to LabID
dtypes: Sparse[float64, nan](635), category(911), float16(31), int64(1)
memory usage: 2.8 GB
None

3095289329


In [18]:
merged_data_all['MaskID'] = merged_data_all['MaskID'].astype('int16')

print(merged_data_all.info())
print()
print(merged_data_all.memory_usage(deep=True).sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923848 entries, 0 to 2923847
Columns: 1578 entries, MaskID to LabID
dtypes: Sparse[float64, nan](635), category(911), float16(31), int16(1)
memory usage: 2.8 GB
None

3077746241


#### turn columns as Sparse[float64] to Sparse[float16]

In [19]:
# Iterate over each column and convert if it's of type Sparse[float64]
for column in merged_data_all.columns:
    if pd.api.types.is_sparse(merged_data_all[column].dtype):
        if merged_data_all[column].dtype.subtype == np.float64:
            merged_data_all[column] = merged_data_all[column].astype('Sparse[float16]')

print(merged_data_all.info())
print()
print(merged_data_all.memory_usage(deep=True).sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923848 entries, 0 to 2923847
Columns: 1578 entries, MaskID to LabID
dtypes: Sparse[float16, nan](635), category(911), float16(31), int16(1)
memory usage: 2.8 GB
None

3024556367


#### turn columns as float16 to category

In [25]:
test_col = merged_data_all['Race_White']

print(test_col.memory_usage(deep=True))
print()

# Convert to float32 or float64 before converting to category
test_col_float32 = test_col.astype('float32')
print(test_col_float32.memory_usage(deep=True))
print()

test_col_category = test_col_float32.astype('category')
print(test_col_category.memory_usage(deep=True))


5847824

11695520

2924076


In [26]:
# Iterate over each column and convert if it's of type float16
for column in merged_data_all.columns:
    if merged_data_all[column].dtype == np.float16:
        print(merged_data_all[column].memory_usage(deep=True))
        #convert to float32 first
        merged_data_all[column] = merged_data_all[column].astype('float32')
        print(merged_data_all[column].memory_usage(deep=True))
        # Convert the column to 'category'
        merged_data_all[column] = merged_data_all[column].astype('category')
        print(merged_data_all[column].memory_usage(deep=True))
        print()

# Display DataFrame info and memory usage
print(merged_data_all.info())
print()
print(merged_data_all.memory_usage(deep=True).sum())


5847824
11695520
2924932

5847824
11695520
2925808

5847824
11695520
5855476

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2925860

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
2924076

5847824
11695520
5862244

5847824
11695520
2924488

5847824
11695520
5851608

5847824
11695520
5851644

5847824
11695520
5851592

5847824
11695520
2924920

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923848 entries, 0 to 2923847
Columns: 1578 entries, MaskID to LabID
dtypes: Sparse[float16, nan](635), category(942), int16(1)
memory usage: 2.

#### Turn SampleMaskID into integer

In [35]:
# SampleMaskID has the biggest memory
# To turn it into interger, we need to check whether all values are starting with 'S', and the rest of values are integer.
column = merged_data_all['SampleMaskID']

print("memory as category: ", column.memory_usage(deep=True))

test_col_obj = test_col.astype('object')
print("memory as object: ",test_col_obj.memory_usage(deep=True))

# Filter out missing values
non_missing_column = column[column.notna()]

# Check if values start with 'S'
starts_with_s = non_missing_column.str.startswith('S')

# Check if the rest part of the values are integers
rest_is_integer = non_missing_column.str.lstrip('S').apply(lambda x: x.isdigit())

# Combine the conditions
all_conditions_met = starts_with_s & rest_is_integer

# Check if all non-missing values in the column meet the conditions
all_values_meet_conditions = all_conditions_met.all()

print("Do all non-missing values start with 'S' and the rest part are integers?", all_values_meet_conditions)

memory as category:  77092158
memory as object:  177606278
Do all non-missing values start with 'S' and the rest part are integers? True


#### No.1 change in column values: remove 'S' for 'SampleMaskID'

In [ ]:
# Strip 's' from the start of each string in the column
merged_data_all['SampleMaskID'] = merged_data_all['SampleMaskID'].str.lstrip('S')
print("original memory usage: ", merged_data_all['SampleMaskID'].memory_usage(deep=True))

# Convert the column to nullable integer type Int16
merged_data_all['SampleMaskID'] = merged_data_all['SampleMaskID'].astype('Int32')
print("memory as Int32: ", merged_data_all['SampleMaskID'].memory_usage(deep=True))

In [48]:
# Display DataFrame info and memory usage
print(merged_data_all.info())
print()
print(merged_data_all.memory_usage(deep=True).sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923848 entries, 0 to 2923847
Columns: 1578 entries, MaskID to LabID
dtypes: Int16(1), Sparse[float16, nan](635), category(941), int16(1)
memory usage: 2.7 GB
None

2880257505


In [ ]:
stripped_column = non_missing_column.str.lstrip('S')

# Convert to int16
stripped_column = stripped_column.astype('int16')

# Handle missing values (if needed)
# If you want to fill missing values with a placeholder (like 0), use:
# df['column_name'] = df['column_name'].fillna(0).astype('int16')
# Or, if you want to keep them as NaN, you can convert them to a float type instead of int16.
# df['column_name'] = df['column_name'].astype('float16')

# Display the updated column info
print(df['column_name'].head())
print(df['column_name'].dtype)


#### turn some category into Sparse[str]

In [31]:
# SexOfRelativeWithT1D14_1 as category in the dataframe with only one value: Male.  
test_col = merged_data_all['SexOfRelativeWithT1D14_1']

print(test_col.memory_usage(deep=True))
print()


test_col_category = test_col.astype('Sparse[str]')
print(test_col_category.memory_usage(deep=True))

2924145

70172521


#### It is an unefficient way to turn string values into sparse even when the column with nearly 100% missing values. 

In [32]:
# SexOfRelativeWithT1D14_1 as category in the dataframe with only one value: Male.  
test_col = merged_data_all['SampleMaskID']

print(test_col.memory_usage(deep=True))
print()


test_col_category = test_col.astype('object')
print(test_col_category.memory_usage(deep=True))

77092158

177606278


In [27]:
import pandas as pd

# Assuming merged_data_all is your DataFrame

def column_memory_usage(df):
    # List to store memory usage and data types
    memory_info = []

    # Iterating through each column to get memory usage and data type
    for col in df.columns:
        memory = df[col].memory_usage(deep=True)
        dtype = df[col].dtype
        memory_info.append({'Column': col, 'Memory Usage (Bytes)': memory, 'Data Type': dtype})

    # Create a DataFrame from the list
    memory_df = pd.DataFrame(memory_info)

    # Sorting by memory usage in descending order
    memory_df = memory_df.sort_values(by='Memory Usage (Bytes)', ascending=False)

    return memory_df

# Call the function and store the result
column_memory_info = column_memory_usage(merged_data_all)

# Display the result
column_memory_info.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1578 entries, 1569 to 1511
Data columns (total 3 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Column                1578 non-null   object
 1   Memory Usage (Bytes)  1578 non-null   int64 
 2   Data Type             1578 non-null   object
dtypes: int64(1), object(2)
memory usage: 49.3+ KB


In [28]:
column_memory_info.to_csv('TN_01_memory_usage_1.csv', index = False)

In [22]:
merged_data_all.to_csv("TN_01_Merged_Data.csv", index = False)

In [23]:
merged_data_all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2923848 entries, 0 to 2923847
Columns: 1578 entries, MaskID to LabID
dtypes: Sparse[float64, nan](635), category(911), float16(31), int64(1)
memory usage: 2.8 GB


In [60]:
series = merged_data_all['Race_White']

print(series.memory_usage(deep=True))

#print(series.astype('float64').memory_usage(deep=True))

#print(series.astype('Sparse').memory_usage(deep=True))

#print(series.astype('int16').memory_usage(deep=True))

#print(series.astype('category').memory_usage(deep=True))

5846108


##### Among 1578 headers of merged_first_part, 1281 of them have over 99% of missing values; 149 of them have 95% - 99% missing values; the rest 148 have less than 95% of missing values. Here are a few suggestions for further analyses.

1). drop those collumes with little little values, such as those with missing values larger than 99%.

2). change to another instance with higher memory.

3). merge some columns together based on their similarity.

#### test of merged dataset

In [2]:
string = "D5OtherTestResultDD"  # change this to the string you're searching for
# Convert the search string to lower case (or upper case)
search_string_lower = string.lower()

# Find columns that contain the search string, case-insensitive
matching_columns = [col for col in transformed_merged_data.columns if search_string_lower in col.lower()]
print("Found ", len(matching_columns), "columns: ", matching_columns)

# Create a DataFrame with the filtered columns and drop rows with only missing values
filtered_df = transformed_merged_data[matching_columns].dropna(how='all')
print("DataFrame with filtered columns (rows with all missing values dropped):")
print()
print(filtered_df.info())
#print(filtered_df.head(30))

# Print top 20 values for each matching column
for col in matching_columns:
    print(f"Top 20 values for column {col}:")
    print(transformed_merged_data[col].value_counts().nlargest(30))
    print()

NameError: name 'transformed_merged_data' is not defined

In [34]:
summary_df = pd.read_csv('merged_first_part_summary.csv')

In [12]:
merged_data_23.memory_usage(deep=True)

Index                                  128
MaskID                            15750456
Date_of_Registration              89150884
Status                            88100924
Date_of_Study_Start               88252854
Date_of_Draw                     123874629
Event_Title                      158197379
Spec_Name                        146762318
SampleMaskID                     120396183
Test_Name                        115208219
Result                           109581057
Result_Type                      113438563
Visit                            124265369
Date_at_Test_Results_Reported    123840259
Date_at_Evaluation               123874419
Date_Received                    121761644
Date_Shipped                     123872389
LabID                             17719263
dtype: int64

In [14]:
merged_data_23['Date_of_Registration'] = pd.to_datetime(merged_data_23['Date_of_Registration'], format='%m/%d/%Y')

In [15]:
merged_data_23.memory_usage(deep=True)

Index                                  128
MaskID                            15750456
Date_of_Registration              15750456
Status                            88100924
Date_of_Study_Start               88252854
Date_of_Draw                     123874629
Event_Title                      158197379
Spec_Name                        146762318
SampleMaskID                     120396183
Test_Name                        115208219
Result                           109581057
Result_Type                      113438563
Visit                            124265369
Date_at_Test_Results_Reported    123840259
Date_at_Evaluation               123874419
Date_Received                    121761644
Date_Shipped                     123872389
LabID                             15750456
dtype: int64

In [17]:
merged_data_23.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1968807 entries, 0 to 1968806
Data columns (total 17 columns):
 #   Column                         Dtype         
---  ------                         -----         
 0   MaskID                         int64         
 1   Date_of_Registration           datetime64[ns]
 2   Status                         object        
 3   Date_of_Study_Start            object        
 4   Date_of_Draw                   object        
 5   Event_Title                    object        
 6   Spec_Name                      object        
 7   SampleMaskID                   object        
 8   Test_Name                      object        
 9   Result                         object        
 10  Result_Type                    object        
 11  Visit                          object        
 12  Date_at_Test_Results_Reported  object        
 13  Date_at_Evaluation             object        
 14  Date_Received                  object        
 15  Date_Shipped   

In [15]:
# Check unique value counts. 
unique_counts = merged_data_23.nunique()
unique_counts

MaskID                           237048
Date_of_Registration               5740
Status                                5
Date_of_Study_Start                5863
Date_of_Draw                       5786
Event_Title                          49
Spec_Name                            22
SampleMaskID                     745974
Test_Name                            37
Result                            10241
Result_Type                           3
Visit                                50
Date_at_Test_Results_Reported      4451
Date_at_Evaluation                 4350
Date_Received                      4162
Date_Shipped                       4673
LabID                                32
dtype: int64

# 1.4 Data query

## 1.4.1 check column names and their values containing a string

In [2]:
# The following code is to load the first part of merged dataset
import pandas as pd
from pandas.core.computation.check import NUMEXPR_INSTALLED
import warnings
warnings.filterwarnings('ignore')

merged_first_part = pd.read_csv('TN_01_merged_data_1.csv')

/home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.0' or newer of 'numexpr' (version '2.7.3' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED


Strings in columns that could be merged:
1. testName: 23
2. 
3. D5OtherTestResultDD, D5OtherTestResultMM, D5OtherTestResultYYYY: 3*23
4. 
5. 
6. 
7. 
8. Heightcm, heightin
9. Weightkg, Weightlbs
10.'TimeOfCollectionHH', 'TimeOfCollectionMM'
11.
12.
13. 

In [2]:
merged_first_part.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 519673 entries, 0 to 519672
Columns: 1563 entries, Unnamed: 0 to SignedConsentFor2017Protocol
dtypes: float64(664), int64(2), object(897)
memory usage: 6.1+ GB


In [25]:
merged_first_part.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 519673 entries, 0 to 519672
Columns: 1563 entries, Unnamed: 0 to SignedConsentFor2017Protocol
dtypes: category(136), float64(664), int64(2), object(761)
memory usage: 5.6+ GB


#### 1) Case insensitive

In [ ]:
string = "D6OtherTestResultUnit"  # change this to the string you're searching for
# Convert the search string to lower case (or upper case)
search_string_lower = string.lower()

# Find columns that contain the search string, case-insensitive
matching_columns = [col for col in merged_first_part.columns if search_string_lower in col.lower()]
print("Found ", len(matching_columns), "columns: ", matching_columns)

# Create a DataFrame with the filtered columns and drop rows with only missing values
filtered_df = merged_first_part[matching_columns].dropna(how='all')
print("DataFrame with filtered columns (rows with all missing values dropped):")
print()
print(filtered_df.info())
#print(filtered_df.head(30))

# Print top 20 values for each matching column
for col in matching_columns:
    print(f"Top 20 values for column {col}:")
    print(merged_first_part[col].value_counts().nlargest(30))
    print()
    

#### 2) Case sensitive

In [ ]:
string = "D5OtherTestResult"  # Change this to the string you're searching for

# Find columns that contain the search string, case-sensitive
matching_columns = [col for col in merged_first_part.columns if string in col]
print("Found ", len(matching_columns), "columns: ", matching_columns)

# Create a DataFrame with the filtered columns and drop rows with only missing values
filtered_df = merged_first_part[matching_columns].dropna(how='all')
print("DataFrame with filtered columns (rows with all missing values dropped):")
print(filtered_df.info())
#print(filtered_df.head(30))

# Print top 20 values for each matching column
for col in matching_columns:
    print(f"Top 20 values for column {col}:")
    print(merged_first_part[col].value_counts().nlargest(20))
    print()


## 1.4.2 check column names and their values containing two strings

Filtered columns:
1. 

In [4]:
# Strings to search for: AND
string1 = ""  # First string
string2 = "heightin"  # Second string, change this to your second string

# Convert the search strings to lower case
search_string1_lower = string1.lower()
search_string2_lower = string2.lower()

# Find columns that contain both search strings, case-insensitive
matching_columns = [col for col in merged_first_part.columns 
                    if search_string1_lower in col.lower() and search_string2_lower in col.lower()]

print("Found ", len(matching_columns), "columns: ", matching_columns)

# Print top 20 values for each matching column
for col in matching_columns:
    print(f"Top 20 values for column {col}:")
    print(merged_first_part[col].value_counts().nlargest(20))
    print()


Found  0 columns:  []


In [ ]:
# Strings to search for: OR
string1 = "heightcm"  # First string
string2 = "heightin"  # Second string, change this to your second string

# Convert the search strings to lower case
search_string1_lower = string1.lower()
search_string2_lower = string2.lower()

# Find columns that contain both search strings, case-insensitive
matching_columns = [col for col in merged_first_part.columns 
                    if search_string1_lower in col.lower() or search_string2_lower in col.lower()]

print("Found ", len(matching_columns), "columns: ", matching_columns)

# Create a DataFrame with the filtered columns and drop rows with only missing values
filtered_df = merged_first_part[matching_columns].dropna(how='all')

# Print top 20 values for each matching column
for col in matching_columns:
    print(f"Top 20 values for column {col}:")
    print(merged_first_part[col].value_counts().nlargest(20))
    print()

In [7]:
filtered_df.describe()

Heightcm    Heightin  Heightcm_dup14  Heightin_dup14  \
count  6322.000000  777.000000      470.000000       22.000000   
mean    145.704470   58.872081      161.463000       59.227273   
std      27.089781   11.184336       19.753296        9.314379   
min       0.000000    0.000000       39.400000       40.900000   
25%     126.500000   51.000000      152.650000       53.035000   
50%     151.000000   62.100000      164.700000       61.000000   
75%     167.000000   67.000000      175.500000       66.400000   
max     219.000000   77.000000      196.170000       72.000000   

       Heightcm_dup15  Heightin_dup15  Heightcm_dup20  Heightin_dup20  \
count    17895.000000     1670.000000       98.000000       26.000000   
mean       155.012744       61.237182      166.091327       65.520385   
std         25.386499       11.930460       15.612072        6.933721   
min          0.000000        0.000000      114.000000       45.000000   
25%        141.000000       56.925000      160.225000       63.075000   
50%        160.500000       64.000000      167.950000       65.750000   
75%        170.700000       68.000000      176.775000       70.922500   
max       1658.500000      185.000000      196.100000       75.000000   

       Heightcm_dup21  Heightin_dup21  
count      150.000000        4.000000  
mean       152.980333       59.100000  
std         20.112152        4.313545  
min         97.200000       53.000000  
25%        138.485000       57.875000  
50%        158.200000       60.200000  
75%        167.002500       61.425000  
max        194.800000       63.000000

## 1.4.3 check column headers in one shared CSV file

In [40]:
import pandas as pd
from pandas.core.computation.check import NUMEXPR_INSTALLED

#file_path = ".././studies/TN_01/TrialNet_01_Data/TN01_ADVERSEEVENTS.CSV"
file_path = "/home/ec2-user/SageMaker/studies/TN_01/TrialNet_01_Data/TN01_nh08_diabonset_current.csv"

df1 = pd.read_csv(file_path)
df1.head()

MaskID    Visit_Dt Visit  \
0  202411  11/06/2018   PRN   
1  202455  10/08/2019   PRN   
2  203313  05/19/2020   PRN   
3  203899  09/15/2020   PRN   
4  207308  02/04/2019   PRN   

                      CriteriaParticipantDiagnosewit  \
0  b.DKA or unequivocally symptomatic AND one dia...   
1  e.Diagnosis made by criteria other than listed...   
2  e.Diagnosis made by criteria other than listed...   
3  b.DKA or unequivocally symptomatic AND one dia...   
4             a.Two consecutive OGTT clinical alerts   

   D1FastingPlasmaGlucDD1_1 D1FastingPlasmaGlucMM1_1  \
0                       NaN                      NaN   
1                      16.0                      Sep   
2                       NaN                      NaN   
3                       NaN                      NaN   
4                       NaN                      NaN   

   D1FastingPlasmaGlucYYYY1_1  D1FastingPlasmaGlucResult1_1  \
0                         NaN                           NaN   
1                      2019.0                          11.6   
2                         NaN                           NaN   
3                         NaN                           NaN   
4                         NaN                           NaN   

  D1FastingPlasmaGlucUnits1_1  D1FastingPlasmaGlucDD2_1  ...  \
0                         NaN                       NaN  ...   
1                      mmol/L                       NaN  ...   
2                         NaN                       NaN  ...   
3                         NaN                       NaN  ...   
4                         NaN                       NaN  ...   

  D5OtherTestResultYYYY14_1  D5OtherTestResultYYYY15_1  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

   D5OtherTestResultYYYY16_1 D5OtherTestResultYYYY17_1  \
0                        NaN                       NaN   
1                        NaN                       NaN   
2                        NaN                       NaN   
3                        NaN                       NaN   
4                        NaN                       NaN   

   D5OtherTestResultYYYY18_1 D5OtherTestResultYYYY19_1  \
0                        NaN                       NaN   
1                        NaN                       NaN   
2                        NaN                       NaN   
3                        NaN                       NaN   
4                        NaN                       NaN   

   D5OtherTestResultYYYY20_1  D5OtherTestResultYYYY21_1  \
0                        NaN                        NaN   
1                        NaN                        NaN   
2                        NaN                        NaN   
3                        NaN                        NaN   
4                        NaN                        NaN   

  D5OtherTestResultYYYY22_1  D5OtherTestResultYYYY23_1  
0                       NaN                        NaN  
1                       NaN                        NaN  
2                       NaN                        NaN  
3                       NaN                        NaN  
4                       NaN                        NaN  

[5 rows x 312 columns]

In [ ]:
string = "MM"  # change this to the string you're searching for
# Convert the search string to lower case (or upper case)
search_string_lower = string.lower()

# Find columns that contain the search string, case-insensitive
matching_columns = [col for col in df1.columns if search_string_lower in col.lower()]
print("Found ", len(matching_columns), "columns: ", matching_columns)


matching_columns2 = ['MaskID', 'Visit_Dt'] + matching_columns
# Create a DataFrame with the filtered columns and drop rows with only missing values
filtered_df = df1[matching_columns2].dropna(how='all', subset=matching_columns)
print("DataFrame with filtered columns (rows with all missing values dropped):")
#print(filtered_df.info())
#print(filtered_df.head(30))

# Print top 20 values for each matching column
for col in matching_columns:
    print(f"Top 20 values for column {col}:")
    print(df1[col].value_counts().nlargest(20))
    print()


In [13]:
# Display the selected rows
filtered_df.head(20)

MaskID    Visit_Dt D5MeterReadingLoc1_1  D5MeterReadingDD1_1  \
0   202411  11/06/2018              In Home                 26.0   
1   202455  10/08/2019            In Clinic                 18.0   
2   203313  05/19/2020              In Home                 23.0   
10  216884  07/29/2019            In Clinic                 23.0   
11  217158  04/16/2019            In Clinic                 26.0   
12  218512  10/26/2020            In Clinic                 17.0   
16  231753  11/25/2019              In Home                 27.0   
22  242474  07/26/2018              In Home                 22.0   
25  247747  04/25/2019              In Home                 24.0   
26  248809  10/13/2020              In Home                 23.0   
27  251390  06/24/2020              In Home                 17.0   
28  254171  08/14/2018              In Home                  1.0   
30  256372  01/23/2019              In Home                 30.0   
32  260645  10/29/2020              Unknown                  NaN   
33  262792  11/29/2018              In Home                  8.0   
34  264001  04/22/2019            In Clinic                 22.0   
36  269436  06/11/2019            In Clinic                  4.0   
37  271758  08/03/2018            In Clinic                 11.0   
38  272788  01/30/2019              In Home                 10.0   
42  289931  08/19/2019              In Home                  9.0   

   D5MeterReadingMM1_1  D5MeterReadingYYYY1_1  D5MeterReadingResult1_1  \
0                  Sep                 2018.0                    160.0   
1                  Sep                 2019.0                     12.5   
2                  Apr                 2020.0                     22.0   
10                 Jul                 2019.0                    325.0   
11                 Mar                 2019.0                     90.0   
12                 Feb                 2020.0                      NaN   
16                 Apr                 2018.0                    138.0   
22                 Aug                 2017.0                    307.0   
25                 Apr                 2019.0                     13.9   
26                 Jun                 2020.0                      7.4   
27                 Jun                 2020.0                     37.0   
28                 Aug                 2018.0                    400.0   
30                 Dec                 2018.0                     11.3   
32                 NaN                    NaN                      NaN   
33                 Nov                 2018.0                    201.0   
34                 Apr                 2019.0                    139.0   
36                 Jun                 2019.0                     11.3   
37                 Jul                 2018.0                      8.8   
38                 Oct                 2018.0                      6.7   
42                 Aug                 2019.0                     29.0   

   D5MeterReadingUnit1_1 D5MeterReadingLoc2_1  D5MeterReadingDD2_1  ...  \
0                  mg/dl                  NaN                  NaN  ...   
1                 mmol/L                  NaN                  NaN  ...   
2                 mmol/L                  NaN                  NaN  ...   
10                 mg/dl                  NaN                  NaN  ...   
11                 mg/dl                  NaN                  NaN  ...   
12                   NaN                  NaN                  NaN  ...   
16                 mg/dl                  NaN                  NaN  ...   
22                 mg/dl            In Clinic                 25.0  ...   
25                mmol/L                  NaN                  NaN  ...   
26                mmol/L                  NaN                  NaN  ...   
27                mmol/L                  NaN                  NaN  ...   
28                 mg/dl                  NaN                  NaN  ...   
30                mmol/L              In Home

In [14]:
# Save the DataFrame to a CSV file
#filtered_df.to_csv('D5MeterReading_TN01_nh08_diabonset_current.csv')

In [6]:
# Filter the dataframe based on the condition
filtered_rows = filtered_df[filtered_df['D5MeterReadingLoc1_1'] == 'Unknown']

# Display the first 20 rows of the filtered dataframe
filtered_rows.head(20)


MaskID    Visit_Dt D5MeterReadingLoc1_1  D5MeterReadingDD1_1  \
32   260645  10/29/2020              Unknown                  NaN   
130  467630  01/14/2020              Unknown                 15.0   
160  518555  06/19/2019              Unknown                  NaN   
167  529847  01/24/2020              Unknown                  NaN   
185  567236  07/18/2018              Unknown                  5.0   
212  600180  08/14/2019              Unknown                  1.0   
258  698893  02/10/2020              Unknown                  NaN   

    D5MeterReadingMM1_1  D5MeterReadingYYYY1_1  D5MeterReadingResult1_1  \
32                  NaN                    NaN                      NaN   
130                 Nov                 2019.0                     17.6   
160                 NaN                    NaN                      NaN   
167                 NaN                    NaN                      NaN   
185                 Jul                 2018.0                    681.0   
212                 Aug                 2019.0                    234.0   
258                 NaN                    NaN                      NaN   

    D5MeterReadingUnit1_1 D5MeterReadingLoc2_1  D5MeterReadingDD2_1  ...  \
32                    NaN                  NaN                  NaN  ...   
130                mmol/L              Unknown                 22.0  ...   
160                   NaN                  NaN                  NaN  ...   
167                   NaN                  NaN                  NaN  ...   
185                 mg/dl                  NaN                  NaN  ...   
212                 mg/dl                  NaN                  NaN  ...   
258                   NaN                  NaN                  NaN  ...   

    D5MeterReadingMM7_1  D5MeterReadingDD5_1  D5MeterReadingDD6_1  \
32                  NaN                  NaN                  NaN   
130                 NaN                  NaN                  NaN   
160                 NaN                  NaN                  NaN   
167                 NaN                  NaN                  NaN   
185                 NaN                  NaN                  NaN   
212                 NaN                  NaN                  NaN   
258                 NaN                  NaN                  NaN   

    D5MeterReadingDD7_1 D5MeterReadingResult5_1  D5MeterReadingResult6_1  \
32                  NaN                     NaN                      NaN   
130                 NaN                     NaN                      NaN   
160                 NaN                     NaN                      NaN   
167                 NaN                     NaN                      NaN   
185                 NaN                     NaN                      NaN   
212                 NaN                     NaN                      NaN   
258                 NaN                     NaN                      NaN   

    D5MeterReadingResult7_1  D5MeterReadingYYYY5_1  D5MeterReadingYYYY6_1  \
32                      NaN                    NaN                    NaN   
130                     NaN                    NaN                    NaN   
160                     NaN                    NaN                    NaN   
167                     NaN                    NaN                    NaN   
185                     NaN                    NaN                    NaN   
212                     NaN                    NaN                    NaN   
258                     NaN                    NaN                    NaN   

    D5MeterReadingYYYY7_1  
32                    NaN  
130                   NaN  
160                   NaN  
167                   NaN  
185                   NaN  
212                   NaN  
258                   NaN  

[7 rows x 44 columns]